### DISCUSSION TITLE
Linux kernel image

### DISCUSSION THREAD

**Discussion by WeetHet** [utmapp/UTM#3577] [UTM Help (macOS)]
Where could I take a linux kernel image for Apple Virtualisation?

**Comment by ghost** [comment]
Yeah, I'd like to know as well. Selecting Apple Virtualization makes the kernel a required field, is this intended? Where should I get a pre-built Kernel image?

**Comment by stan-kondrat** [comment]
I found some Information, but I didn't succeed to launch 

https://developer.apple.com/documentation/virtualization/running_linux_in_a_virtual_machine

> `You may obtain a kernel image and the corresponding initial RAM disk image for a given release of the Fedora Linux distribution from https://download.fedoraproject.org/pub/fedora/linux/releases/<release>/Everything/<architecture>/os/images/pxeboot, where <release> is the Fedora release number and <architecture> is x86_64 for Intel Macs and aarch64 for Apple silicon Macs.`

**Comment by 19h** [comment]
You can download something like ubuntu, extract initrd (ramdisk) and vmlinuz from casper/initrd and caspter/vmlinuz, respectively. Then use vmlinuz as the kernel image and initrd as initial ramdisk, select the ubuntu distribution iso as boot image. It should run just fine.

**Reply by pkpfr** [reply]
Unfortunately. this is easier said than done.  The kernel and image need to be uncompressed.  The scripts for decompressing kernels don't work on MacOS.  It would be great if there was a straightforward tutorial on the steps to uncompress these files on an Apple Silicon device.

**Reply by ktprograms** [reply]
I'm not sure what scripts you're referring to, but if it's just a gzip (you can check with `file <path_to_kernel>` in a Terminal), then just rename it adding a `.gz` extension then run `gunzip <kernel.gz>` in a Terminal)

**Reply by pkpfr** [reply]
That's may work for some distributions, but not for Debian or Ubuntu kernels that contain additional header data before the compressed content and do not use gzip compression. 

The script I refer to is https://github.com/torvalds/linux/blob/master/scripts/extract-vmlinux.

**Reply by ktprograms** [reply]
The only part of the script I can see that might not work is the readelf, and you can always manually try the `tr`/`grep` stuff to uncompress it.

(Sorry I sent this as a new answer the first time)

**Reply by pkpfr** [reply]
Yes, I've tried searching manually in both MacOS and a virtualised version of Debian running on the M1.  Unfortunately I'm not finding any of the compression headers in Debian's latest vmlinuz-5.10.0-15-arm64 file. I'll keep plugging away, but all paths currently point to Linus' 3 year old script.

**Reply by 19h** [reply]
> Unfortunately. this is easier said than done. The kernel and image need to be uncompressed. The scripts for decompressing kernels don't work on MacOS. It would be great if there was a straightforward tutorial on the steps to uncompress these files on an Apple Silicon device.

I don’t understand this. I literally am running Ubuntu on my M1 Max MBP (macOS Ventura) using these exact files. You can use `7z x ubuntu.iso` to decompress the image and then use the files as specified above.

**Reply by ballo** [reply]
I tried to get a debian based distro running (parrotos) and failed miserably. I'm only trying this because QEMU USB passthrough support is broken. I tried using the initrd.img and vmlinuz files from an existing install. The utter lack of documentation or specifications as to what files to use in what format is frustrating.

I have zero interest in using Ubuntu (which is honestly more bloated than Microsoft Windows these days)

**Reply by ballo** [reply]
file says vmlinuz is a Linux kernel

**Reply by ktprograms** [reply]
Are both the initrd and vmlinuz **not** gzipped?

**Reply by ballo** [reply]
Correct

**Comment by pkpfr** [comment]
When I try, I just get a black screen on boot.  As mentioned, a couple of paragraphs in the readme on the exact steps to take to get The latest Ubuntu running under Apple Virtualisation would be incredibly valuable.

**Reply by ktprograms** [reply]
Are you using x86_64? IIRC the x86_64 kernels aren't compressed whole, instead there's a decompressor built in.

**Comment by pkpfr** [comment]
No, I’m using an ARM64 Apple M1.

**Reply by pkpfr** [reply]
The ARM64 Ubuntu files are compressed with gzip (vmlinuz) and zstd (initrd).  I have uncompressed both files and but it still starts with a blank window.  There's another user with a similar problem, so I have added the console output to that issue.  https://github.com/utmapp/UTM/issues/4004#issuecomment-1154777019.

**Comment by vincent-olivert-riera** [comment]
> The ARM64 Ubuntu files are compressed with gzip (vmlinuz) and zstd (initrd). I have uncompressed both files and but it still starts with a blank window. There's another user with a similar problem, so I have added the console output to that issue. [#4004 (comment)](https://github.com/utmapp/UTM/issues/4004#issuecomment-1154777019).

Same problem here. I'm using the kernel and initrd (I have uncompressed it with `gzip -d`) from Debian:
```
Debian % file *
initrd: ASCII cpio archive (SVR4 with no CRC)
linux:  Linux kernel ARM64 boot executable Image, little-endian, 4K pages
```

The files were downloaded from here:
- https://deb.debian.org/debian/dists/bullseye/main/installer-arm64/current/images/netboot/debian-installer/arm64/linux
- https://deb.debian.org/debian/dists/bullseye/main/installer-arm64/current/images/netboot/debian-installer/arm64/initrd.gz

**Comment by sscarduzio** [comment]
Did anyone have luck booting any Linux distro in UTM on M1 with Apple Virtualization enabled?

**Comment by LXY1226** [comment]
emm... There has a correct configuration
kernel Image: [ubuntu-21.04-server-cloudimg-arm64-vmlinuz-generic](https://cloud-images.ubuntu.com/releases/hirsute/release/unpacked/ubuntu-21.04-server-cloudimg-arm64-vmlinuz-generic)
initrd: [ubuntu-21.04-server-cloudimg-arm64-initrd-generic](https://cloud-images.ubuntu.com/releases/hirsute/release/unpacked/ubuntu-21.04-server-cloudimg-arm64-initrd-generic)
kernel param: `console=hvc0`

after downloaded, `mv ubuntu-21.04-server-cloudimg-arm64-vmlinuz-generic ubuntu-21.04-server-cloudimg-arm64-vmlinuz-generic.gz && gunzip ubuntu-21.04-server-cloudimg-arm64-vmlinuz-generic.gz`, then it should works.

also alpine linux is working too.
iso: [alpine-virt-3.16.0-aarch64.iso](https://mirrors.tuna.tsinghua.edu.cn/alpine/v3.16/releases/aarch64/alpine-virt-3.16.0-aarch64.iso) # I'm in China, so change it to your closer mirror
unpack it, gunzip boot/vmlinux, and it should work with same kernel param

**Reply by tleerai** [reply]
Receiving error:

`No root device specified. Boot arguments must include a root= parameter`

**Reply by Sitin** [reply]
For 22.04 here is the root folder for downloads [https://cloud-images.ubuntu.com/releases/22.04/release/unpacked/](https://cloud-images.ubuntu.com/releases/22.04/release/unpacked/)

**Reply by Sitin** [reply]
This setup gives me a minimal Ubuntu OS. At the moment stuck on booting with iso image with Ubuntu Server. Any suggestions?

**Comment by Sitin** [comment]
Kernel and ramdisk for 22.04 can be found here: [https://cloud-images.ubuntu.com/releases/22.04/release/unpacked/](https://cloud-images.ubuntu.com/releases/22.04/release/unpacked/).

```shell
mkdir -p ~/Downloads/ubuntu-utm
cd ~/Downloads/ubuntu-utm

# For kerken
curl https://cloud-images.ubuntu.com/releases/22.04/release/unpacked/ubuntu-22.04-server-cloudimg-arm64-vmlinuz-generic -o ubuntu-22.04-server-cloudimg-arm64-vmlinuz-generic.gz
gunzip ubuntu-22.04-server-cloudimg-arm64-vmlinuz-generic.gz

# For ramdisk
curl https://cloud-images.ubuntu.com/releases/22.04/release/unpacked/ubuntu-22.04-server-cloudimg-arm64-initrd-generic -o ubuntu-22.04-server-cloudimg-arm64-initrd-generic
```

Boot params: `console=hvc0 rd.break=initqueue`

I was able to boot "something". Will create a gist and HOWTO one there will be any success.

**Comment by testn** [comment]
Did anyone find a working instruction for this?

**Comment by rebooting** [comment]
thanks folks for the hints on how to set up the VM!

using @LXY1226 's instructions, I've managed to boot the Ubuntu cloud image.
Main points:

- you need to figure out what is your image's device name ( /dev/vda , or /dev/vdb for example, usually just /dev/vda if you only have a disk image configured)
- the boot uses raw disk, you need to use the raw image ( e.g. download and untar this [https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-arm64.tar.gz](https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-arm64.tar.gz)
- boot params ```console=hvc0 root=/dev/vda``` or ```console=hvc0 root=/dev/vdb``` depending on your device name and order

that's the least number of steps to see it boot up to the OS.

You will need a cloud init iso to configure the user for login ( and update packages)


I've documented it here, let me know if it's useful.

https://workinginavirtualspace.blogspot.com/2022/11/using-umt-to-create-virtual-machines.html

**Reply by Sheorath** [reply]
Works very well, documentation could be a bit better

For People who are lost with the boot.iso (seed.iso):
https://www.how2shout.com/linux/install-ubuntu-20-04-22-04-cloud-image-minimal-on-virtualbox/#3_Download_SeedISO_file

**Comment by chrispreee** [comment]
Enormous thanks LXY1226 and rebooting

Has anyone got `ubuntu-desktop` to work on Apple Hypervisor? It installed but boots into CLI (I presume due to a graphics issue).

**Reply by ktprograms** [reply]
GUI Linux with Apple Virtualization.framework is only available on macOS 13. With macOS 13, Apple Virtualization.framework also supports EFI/ISO boot, so you don't need to extract the kernel image.

**Comment by XavierFilipe** [comment]
I'm using fedora 38 arm64, and i would like to compile upstream kernel linux (from source) to use on UTM (M1 Pro), I tried this:

```
make localmodconfig

make LLVM=1 -j 4

sudo make modules_install
sudo make install
sudo grub2-mkconfig -o /boot/grub2/grub.cfg
```

but it doesn't work, it crashes on startup, even if you enable loglevel=7 it doesn't show the logs.

<img width="619" alt="image" src="https://github.com/user-attachments/assets/6a39fd96-9ecd-4ae8-850f-c684fc6a1d26" />


**What kernel flags are needed to run Linux on UTM?**
I defined CONFIG_EFI_STUB=y in my .config.

**Which compression method should be used?**

**Reply by XavierFilipe** [reply]
https://docs.getutm.app/guest-support/linux/

If you are building your own kernel, make sure to include the following options:

```
CONFIG_VIRTIO
CONFIG_VIRTIO_RING
CONFIG_VIRTIO_PCI
CONFIG_VIRTIO_BALLOON
CONFIG_VIRTIO_BLK for storage devices
CONFIG_VIRTIO_CONSOLE for console devices
CONFIG_VIRTIO_NET for networking
CONFIG_DRM_VIRTIO_GPU for graphical output
CONFIG_NET_9P, CONFIG_NET_9P_VIRTIO, CONFIG_9P_FS, CONFIG_9P_FS_POSIX_ACL for VirtFS directory sharing
CONFIG_VIRTIO_FS for VirtioFS directory sharing
```

this worked for me.
